src/main/java/org/jboss/modules/ModuleLoggerFinder.java

```diff
@@ -34,64 +43,175 @@
  * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
  */
 public final class ModuleLoggerFinder extends LoggerFinder {
+    private static final LoggerFinder queueingFinder = new LoggerFinder() {
+        @Override
+        public System.Logger getLogger(final String name, final java.lang.Module module) {
+            return new DelegatingSystemLogger(new QueueingSystemLogger(name, module));
+        }
+    };
 
-    private final Supplier<LoggerFinder> defaultFinder = () -> new LoggerFinder() {
+    private static final Supplier<LoggerFinder> defaultFinder = () -> new LoggerFinder() {
         @Override
         public System.Logger getLogger(final String name, final java.lang.Module module) {
             return new JulSystemLogger(Logger.getLogger(name));
         }
     };
-    private static volatile ClassLoader classLoader;
-    private volatile LoggerFinder finder;
-
-    public ModuleLoggerFinder() {
-    }
 
+    private static final Map<String, System.Logger> loggers = new ConcurrentHashMap<>();
+    private static final ReentrantLock lock = new ReentrantLock();
+    private static final Deque<SystemLogRecord> messages = new LinkedBlockingDeque<>();
+    private static volatile boolean activated = false;
+    // Guarded by lock
+    private static LoggerFinder finder;
```

Comment:
Although it is guarded by lock it needs to be declared volatile to ensure visibility according to Java Memory model

Reply:
It's only accessed for read outside the lock, however it should only ever happen once `activated` is `true` which is set within the lock and it, itself is `volatile`. We could add it, I just didn't think the extra read overhead was worth it.